addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java

```diff
@@ -0,0 +1,255 @@
+import org.openhab.binding.paradoxalarm.internal.communication.ParadoxCommunicatorFactory;
+import org.openhab.binding.paradoxalarm.internal.exceptions.ParadoxBindingException;
+import org.openhab.binding.paradoxalarm.internal.model.PanelType;
+import org.openhab.binding.paradoxalarm.internal.model.RawStructuredDataCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ParadoxIP150BridgeHandler} This is the handler that takes care of communication to/from Paradox alarm
+ * system.
+ *
+ * @author Konstantin_Polihronov - Initial contribution
+ */
+@NonNullByDefault
+public class ParadoxIP150BridgeHandler extends BaseBridgeHandler {
+
+    private static final String RESET = "RESET";
+
+    private static final int INITIAL_SCHEDULE_DELAY = 15;
+    private static final int FAILED_READ_ATTEMPTS_TRESHOLD = 10;
+
+    private final Logger logger = LoggerFactory.getLogger(ParadoxIP150BridgeHandler.class);
+
+    private static IParadoxCommunicator communicator;
+    private static ParadoxIP150BridgeConfiguration config;
+    private @Nullable ScheduledFuture<?> refreshCacheUpdateSchedule;
+    private int failedReadAttempts;
+
+    public ParadoxIP150BridgeHandler(Bridge bridge) throws Exception {
+        super(bridge);
+    }
+
+    private synchronized IParadoxCommunicator initializeCommunicator() throws Exception {
+        String ipAddress = config.getIpAddress();
+        int tcpPort = config.getPort();
+        String ip150Password = config.getIp150Password();
+        String pcPassword = config.getPcPassword();
+
+        logger.info("Phase1 - Identify communicator");
+        IParadoxGenericCommunicator initialCommunicator = new GenericCommunicator(ipAddress, tcpPort, ip150Password,
+                pcPassword);
+        byte[] panelInfoBytes = initialCommunicator.getPanelInfoBytes();
+
+        PanelType panelType = PanelType.parsePanelType(panelInfoBytes);
+        logger.info("Found {} panel type.", panelType);
+        initialCommunicator.close();
+
+        // If not detected properly, use the value from config
+        String panelTypeStr;
+        if (panelType != PanelType.UNKNOWN) {
+            panelTypeStr = panelType.name();
+        } else {
+            panelTypeStr = config.getPanelType();
+        }
+
+        logger.info("Phase2 - Creating communicator for panel {}", panelType);
+        ParadoxCommunicatorFactory factory = new ParadoxCommunicatorFactory(ipAddress, tcpPort, ip150Password,
+                pcPassword);
+        return factory.createCommunicator(panelTypeStr);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Start initializing!");
+        updateStatus(ThingStatus.UNKNOWN);
+        try {
+            logger.info("Starting creation of communicator handler");
+            config = getConfigAs(ParadoxIP150BridgeConfiguration.class);
+            communicator = initializeCommunicator();
```

Comment:
Does that mean that I can keep the bridge OFFLINE until the scheduled task is being executed successfully? Is that OK? I guess you advice me to just read the config, set the fields and continue offline until the execution of connect to the system happens...

Reply:
Yes that is how it done in most cases. Set the bridge offline or unknown and when the actual device information is received set the state based on that information.

Reply:
Just want to let you all know that I'm currently refactoring the whole lifecycle and execution of handlers to use async methods. Need more decoupling though so it will take time before the actual commit...

Reply:
Everything should have been fixed now. Looks OK to me. Please review.